pkg/lib/hooks.js

```diff
@@ -0,0 +1,291 @@
+ * dependencies, such as options for a Cockpit channel.  However,
+ * React will compare dependency values with Object.is, and would run
+ * the effect hook too often.  In the example above, the "options"
+ * variable of Component is a different object on each render
+ * according to Object.is, but we only want to open a new channel when
+ * the value of a field such as "superuser" or "host" has actually
+ * changed.
+ *
+ * A call to useDeepEqualMemo will return some object that is deeply
+ * equal to its argument, and it will continue to return the same
+ * object (according to Object.is) until the parameter is not deeply
+ * equal to it anymore.
+ *
+ * For the example, this means that "memo_options" will always be the
+ * very same object, and the effect hook is only run once.  If we
+ * would use "options" directly as a dependency of the effect hook,
+ * the channel would be closed and opened on every render. This is
+ * very inefficient, doesn't give the asynchronous channel time to do
+ * its job, and will also lead to infinite loops when events on the
+ * channel cause re-renders (which in turn will run the effect hook
+ * again, which will cause a new event, ...).
+ */
+
+export function useDeepEqualMemo(value) {
+    const ref = useRef(value);
+    if (!deep_equal(ref.current, value))
+        ref.current = value;
+    return ref.current;
+}
+
+/* - useFile(path, options)
+ *
+ * function Component() {
+ *   const content = useFile("/etc/hostname", { superuser: "try" });
+ *   const [content, error] = useFileWithError("/etc/hostname", { superuser: "try" });
+ *
+ *   ...
+ * }
+ *
+ * The "path" and "options" parameters are passed unchanged to
+ * cockpit.file().  Thus, if you need to parse the content of the
+ * file, the best way to do that is via the "syntax" option.
+ *
+ * The "content" variable will reflect the content of the file
+ * "/etc/hostname". When the file changes on disk, the component will
+ * be re-rendered with the new content.
+ *
+ * When the file does not exist or there has been some error reading
+ * it, "content" will be false.
+ *
+ * The "error" variable will contain any errors encountered while
+ * reading the file.  It is false when there are no errors.
+ *
+ * When the file does not exist, "error" will be false.
+ *
+ * The "content" and "error" variables will be null until the file has
+ * been read for the first time.
+ *
+ * useFile and useFileWithError are pretty much the same. useFile will
+ * hide the exact error from the caller, which makes it slightly
+ * cleaner to use when the exact error is not part of the UI. In the
+ * case of error, useFile will log that error to the console.
+ */
+
+export function useFileWithError(path, options, hook_options) {
+    const [content_and_error, setContentAndError] = useState([null, null]);
+    const memo_options = useDeepEqualMemo(options);
+    const memo_hook_options = useDeepEqualMemo(hook_options);
```

Comment:
OOI, is that for explicitly supporting *changing* the options for an existing instantiated component/hook? In practice, these should always be constant, no?

Reply:
Correct. I have done this mostly to make eslint happy, it would otherwise complain about a missing dependency for `useEffect`. And since eslint is technically correct, I didn't want to override it and rather took the small performance hit of also memoizing the hook options.